tdrs-backend/tdpservice/parsers/case_consistency_validator.py

```diff
@@ -0,0 +1,125 @@
+        self.total_cases_cached = 0
+        self.total_cases_validated = 0
+
+    def clear_errors(self):
+        """Reset generated errors."""
+        self.generated_errors = []
+
+    def get_generated_errors(self):
+        """Return all errors generated for the current validated case."""
+        if self.has_validated:
+            return self.generated_errors
+        return []
+
+    def add_record(self, record, schema, case_has_errors):
+        """Add record to cache and validate if new case is detected."""
+        if self.case_is_section_one_or_two:
+            if record.CASE_NUMBER != self.current_case and self.current_case is not None:
+                self.validate()
+                self.record_schema_pairs = [(record, schema)]
+                self.case_has_errors = case_has_errors
+            else:
+                self.case_has_errors = self.case_has_errors if self.case_has_errors else case_has_errors
+                self.record_schema_pairs.append((record, schema))
+                self.has_validated = False
+            self.current_case = record.CASE_NUMBER
+
+    def validate(self):
+        """Perform category four validation on all cached records."""
+        num_errors = 0
+        if self.case_is_section_one_or_two:
+            self.total_cases_cached += 1
+            if not self.case_has_errors:
+                self.total_cases_validated += 1
+                self.has_validated = True
+                logger.debug(f"Attempting to execute Cat4 validation for case: {self.current_case}.")
+                if self.program_type == "TAN" and self.section == "A" and "state_fips" in self.header:
+                    num_errors += self.__validate_tanf_s1_case(num_errors)
+                elif self.program_type == "TAN" and self.section == "C" and "state_fips" in self.header:
+                    num_errors += self.__validate_tanf_s2_case(num_errors)
+                elif self.program_type == "TAN" and self.section == "A" and "tribe_code" in self.header:
+                    num_errors += self.__validate_tribal_tanf_s1_case(num_errors)
+                elif self.program_type == "TAN" and self.section == "C" and "tribe_code" in self.header:
+                    num_errors += self.__validate_tribal_tanf_s2_case(num_errors)
+                elif self.program_type == "SSP" and self.section == "A":
+                    num_errors += self.__validate_ssp_s1_case(num_errors)
+                elif self.program_type == "SSP" and self.section == "C":
+                    num_errors += self.__validate_ssp_s2_case(num_errors)
+                else:
+                    self.total_cases_validated -= 1
+                    logger.warn(f"Case: {self.current_case} has no errors but has either an incorrect program type: "
+                                f"{self.program_type} or an incorrect section: {self.section}. No validation occurred.")
+                    self.has_validated = False
+            else:
```

Comment:
This is ok for a MVP but could definitely be improved in a polished version. Thoughts?

Reply:
See Jan's PR for an update to this.

Reply:
@andrew-jameson